woocommerce/woocommerce
# Adding streaming support through Jetpack API

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

This PR switches over to using the Jetpack `jetpack-ai-query` streaming endpoint for product descriptions, and implementing that within the editor. It utilizes some utilities copied from the Jetpack AI block, and I created the `useCompletion` hook to simplify it's usage.

This also removes the need for a fair amount of code, perhaps including our own REST endpoints but I've left those just in case @nima-karimi 's use-case requires it. 

Closes #38413 .

### Screenshots
![image](https://github.com/woocommerce/woocommerce/assets/444632/2ecf7f16-c521-42c0-b17d-14185cf9b5e9)




### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. I've been using this with a WoA dev site ([guide](https://explorersp2.wordpress.com/2021/11/08/wordpress-on-atomic-development-workflow/)). Following that guide, but with [this modified script](https://gist.github.com/joelclimbsthings/d0144f525e165d4943650feeec09c9a4).
2. Once you've copied the plugin over, ensure the `Woo AI` plugin is activated.
3. Go to Products -> Add new
4. You should see the "Write it for me" button as pictured above.
5. Observe that it's disabled until you enter a Product title of ~20 characters.
6. After entering a title, it should be enabled.
7. After clicking the button, you should see the familiar streaming behavior within the description editor creating your description.
8. The button with change to "Stop writing"
9. Ensure the the completion finishes without an issue.
10. Then try again, and hit "Stop writing" to ensure that the streaming stops as expected.
